sql/src/main/java/io/crate/execution/engine/aggregation/impl/CollectSetAggregation.java

```diff
@@ -72,26 +74,51 @@ public FunctionInfo info() {
     }
 
     @Override
-    public Set<Object> iterate(RamAccountingContext ramAccountingContext, Set<Object> state, Input... args) throws CircuitBreakingException {
+    public Map<Object, int[]> iterate(RamAccountingContext ramAccountingContext,
+                                      Map<Object, int[]> state,
+                                      Input... args) throws CircuitBreakingException {
         Object value = args[0].value();
         if (value == null) {
             return state;
         }
-        if (state.add(value)) {
-            ramAccountingContext.addBytes(
-                RamAccountingContext.roundUp(innerTypeEstimator.estimateSize(value) + 36L) // values size + 32 bytes for entry, 4 bytes for increased capacity
-            );
-        }
+
+        upsertOccurenceForValue(value, 1, state, ramAccountingContext);
+
         return state;
     }
 
+    private void upsertOccurenceForValue(Object value,
+                                         int occurrenceIncrement,
+                                         Map<Object, int[]> state,
+                                         RamAccountingContext ramAccountingContext) {
+        state.compute(value, (k, v) -> {
+            if (v == null) {
+                ramAccountingContext.addBytes(
+                    // values size + 32 bytes for entry, 4 bytes for increased capacity, 4 bytes for the count integer
+                    RamAccountingContext.roundUp(innerTypeEstimator.estimateSize(value) + 40L)
+                );
+                if (occurrenceIncrement == 1) {
+                    return ONE_OCCURENCE;
+                } else {
+                    return new int[]{occurrenceIncrement};
+                }
+            } else if (v == ONE_OCCURENCE) {
+                // paramount we create a new occurence container at this stage so we don't update the shared ONE_OCC one
+                return new int[]{occurrenceIncrement + 1};
+            } else {
+                v[0] += occurrenceIncrement;
```

Comment:
why do we track number of occurrences?

Reply:
in the removable cumulative mode we need not remove the value from the set if it occurred multiple times 

eg. for windowing mode CR - Unbounded Following for an unordered window 1, 2, 2, 2, 3 the window frames will be {1, 2, 2, 2, 3}, {2, 2, 2, 3}, {2, 2, 3},  {2, 3} and {3}. The value for collect_set for this case needs to be [1, 2, 3], [2, 3], [2, 3], [2, 3], [3] (without keeping track of occrrences the 3rd frame ( {2, 2, 3} ) will yiled a set of [3] which is incorrect

Had we not have this count

Reply:
can we please document that somehow? this requirement of a window function logic is not visible inside the aggregation. In general, I think the `removable cumulative mode`  statement is not very verbose when reading it outside of any window function scope (commit message, comments etc)